src/shims/unix/fs.rs

```diff
@@ -1086,8 +1088,18 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                 let name_bytes = name.as_encoded_bytes();
                 let name_len = u64::try_from(name_bytes.len()).unwrap();
 
-                let dirent64_layout = this.libc_ty_layout("dirent64");
-                let d_name_offset = dirent64_layout.fields.offset(4 /* d_name */).bytes();
+                let is_linux = matches!(&*this.tcx.sess.target.os, "linux");
+
+                let dirent64_layout = if is_linux {
+                    this.libc_ty_layout("dirent64")
+                } else {
+                    this.libc_ty_layout("dirent")
+                };
+                let d_name_offset = if is_linux {
+                    dirent64_layout.fields.offset(4 /* d_name */).bytes()
+                } else {
+                    dirent64_layout.fields.offset(3 /* d_name */).bytes()
+                };
```

Comment:
This has to always be the last field, right? Otherwise our allocation logic makes no sense. So please make it pick the last field here rather than hard-coding an index.